pubsublite/types.go

```diff
@@ -0,0 +1,140 @@
+// should have the format: `projects/{project}/locations/{zone}/topics/{id}`.
+func ParseTopicPath(input string) (TopicPath, error) {
+	parts := topicPathRE.FindStringSubmatch(input)
+	if len(parts) < 4 {
+		return TopicPath{}, fmt.Errorf("pubsublite: invalid topic path %q", input)
+	}
+	return TopicPath{Project: parts[1], Zone: parts[2], TopicID: parts[3]}, nil
+}
+
+// SubscriptionPath stores the full path of a Google Pub/Sub Lite subscription.
+// See https://cloud.google.com/pubsub/lite/docs/subscriptions for more
+// information.
+type SubscriptionPath struct {
+	// A Google Cloud project. The project ID (e.g. "my-project") or the project
+	// number (e.g. 987654321) can be provided.
+	Project string
+
+	// A Google Cloud zone. An example zone is "us-central1-a".
+	// See https://cloud.google.com/pubsub/lite/docs/locations for the list of
+	// zones where Google Pub/Sub Lite is available.
+	Zone string
+
+	// The ID of the Google Pub/Sub Lite subscription, for example
+	// "my-subscription-name".
+	SubscriptionID string
+}
+
+func (s SubscriptionPath) String() string {
+	return fmt.Sprintf("projects/%s/locations/%s/subscriptions/%s", s.Project, s.Zone, s.SubscriptionID)
+}
+
+// Location returns the path of the parent location.
+func (s SubscriptionPath) Location() LocationPath {
+	return LocationPath{Project: s.Project, Zone: s.Zone}
+}
+
+var subsPathRE = regexp.MustCompile(`^projects/([^/]+)/locations/([^/]+)/subscriptions/([^/]+)$`)
+
+// ParseSubscriptionPath parses the full path of a Google Pub/Sub Lite
+// subscription, which should have the format:
+// `projects/{project}/locations/{zone}/subscriptions/{id}`.
+func ParseSubscriptionPath(input string) (SubscriptionPath, error) {
+	parts := subsPathRE.FindStringSubmatch(input)
+	if len(parts) < 4 {
+		return SubscriptionPath{}, fmt.Errorf("pubsublite: invalid subscription path %q", input)
+	}
+	return SubscriptionPath{Project: parts[1], Zone: parts[2], SubscriptionID: parts[3]}, nil
+}
+
+// ValidateZone verifies that the `input` string has the format of a valid
+// Google Cloud zone. An example zone is "europe-west1-b".
+// See https://cloud.google.com/compute/docs/regions-zones for more information.
+func ValidateZone(input string) error {
+	parts := strings.Split(input, "-")
+	if len(parts) != 3 {
+		return fmt.Errorf("pubsublite: invalid zone %q", input)
+	}
+	return nil
+}
+
+// ValidateRegion verifies that the `input` string has the format of a valid
+// Google Cloud region. An example region is "europe-west1".
+// See https://cloud.google.com/compute/docs/regions-zones for more information.
+func ValidateRegion(input string) error {
+	parts := strings.Split(input, "-")
+	if len(parts) != 2 {
+		return fmt.Errorf("pubsublite: invalid region %q", input)
+	}
+	return nil
+}
+
+// ZoneToRegion returns the region that the given zone is in.
+func ZoneToRegion(zone string) string {
+	return zone[0:strings.LastIndex(zone, "-")]
```

Comment:
Since this is a publicly exposed function, we might want to consider bounds checking in the case the user passes in an invalid string (perhaps by a check to `ValidateZone`)?

Reply:
Done, good catch. I originally had this as a function of an already validated CloudZone, but correctness is not guaranteed now.